### PR TITLE
Add Data.List.NonEmpty.Extra

### LINKED ISSUE #44: Add Data.List.NonEmpty.Extra?
I use `NonEmpty` a lot and these are the functions I find useful:

```haskell
-- | /O(n)/. Append an element to a list.
(|:) :: [a] -> a -> NonEmpty a
(|:) xs x = foldr cons (pure x) xs

-- | Append a list to a non-empty list.
appendl :: NonEmpty a -> [a] -> NonEmpty a
appendl (x :| xs) l = x :| (xs ++ l)

-- | Append a non-empty list to a list.
appendr :: [a] -> NonEmpty a -> NonEmpty a
appendr l nel = foldr cons nel l
```

There are certainly other useful functions for `NonEmpty` (e.g., the counterparts of many functions in `Data.List.Extra`), but I find these particularly useful since they obviate the need to use the partial `fromList`. So hopefully this is a reasonable starting point.

### PR DESCRIPTION
**Pull request by zliu41**
Closes #44 

There's a problem with generating `Extra.hs`, since there are name clashes between `Data.List.NonEmpty` and `Data.List.Extra`, such as `cons`. Is there an easy workaround? Maybe remove it?

### CODE REVIEW

**Comment by ndmitchell** [comment]
I suggest changing the generator so that it imports only a subset of NonEmpty.

Note that you need to check in the generated code.

**Comment by zliu41** [comment]
Updated generator.

File: Generate.hs [hunk 1]
```diff
@@ -37,11 +37,11 @@ main = do
         ,"module Extra {-# DEPRECATED \"This module is provided as documentation of all new functions, you should import the more specific modules directly.\" #-} ("] ++
         concat [ ["    -- * " ++ mod
                  ,"    -- | Extra functions available in @" ++ show mod ++ "@."
-                 ,"    " ++ unwords (map (++",") funs)]
+                 ,"    " ++ (unwords . map (++",") . filter (notHidden mod) $ funs)]
```
Comment (ndmitchell) [inline Generate.hs]:
```suggestion
                 ,"    " ++ unwords (map (++",") $ filter (notHidden mod) funs)]
```

**Comment by zliu41** [comment]
> Looks great to me - you happy for me to merge?

Yeah, but I still need to make the build pass for GHC <= 8.2 though

File: extra.cabal [hunk 2]
```diff
@@ -36,6 +36,7 @@ library
         filepath,
         process,
         clock >= 0.7,
+        semigroups,
```
Comment (zliu41) [inline extra.cabal]: This is for GHC 7.10 since `NonEmpty` wasn't added to `base` then.

Comment (ndmitchell) [reply extra.cabal]: Can we get a comment to that effect, so I know when I upgrade to GHC 8 only I can delete this part?

**Comment by ndmitchell** [comment]
Thanks for the PR! Happy to make a release now, or hold off a bit if you want to get more changes in before the release.

**Comment by zliu41** [comment]
Thanks for the quick feedback. I don't need an immediate release; the new module probably deserves some more functions and I can add them when I get a chance.

**Comment by ndmitchell** [comment]
Cool, will wait for you to say it should be released.

**Pull request merged**